File: Data Structures & Algorithms/climbing-stairs/submission-1.cpp
class Solution {
public:
    int climbStairs(int n) {
        vector<int> dp(n+1);
        if(n<=1){
            return 1;
        }
        dp[0] = 1;
        dp[1]= 1;
        int i = 2;
        while(i<=n){
            dp[i] = dp[i-1]+dp[i-2];
            i++;
        }
        return dp[n];
    }
};

File: Data Structures & Algorithms/course-schedule-ii/submission-1.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = {i:[] for i in range(numCourses)}

        for crs,pre in prerequisites:
            adj[crs].append(pre)
        
        visit = set()
        cycle = set()
        result = []

        def dfs(crs):
            if crs in visit:
                return True
            if crs in cycle:
                return False
            
            cycle.add(crs)
            for n in adj[crs]:
                if not dfs(n):
                    return False
            cycle.remove(crs)
            visit.add(crs)
            result.append(crs)
            return True
        for c in range(numCourses):
            if not dfs(c):
                return []
        return result


